### PR TITLE
#1168 Add support for array parameters on the SQLFilter

### PR DESCRIPTION
**Pull request by mnogales**

### CODE REVIEW

**Comment by holtkamp** [comment]
This resolves #2624

**Comment by zeroedin-bill** [comment]
This adds a few new ways in which the SQLFilter can fail horribly. If objects are used as parameters, and the objects are later changed without calling setParameter again, this would break query caching in hard to debug ways.

**Comment by holtkamp** [comment]
@zeroedin-bill, and that would not be possible when only one parameter was used instead of an array of parameters?

File: lib/Doctrine/ORM/Query/Filter/SQLFilter.php [hunk 1]
```diff
@@ -72,7 +73,11 @@
     final public function setParameter($name, $value, $type = null)
     {
         if (null === $type) {
-            $type = ParameterTypeInferer::inferType($value);
+            if (is_array($type)) {
```
Comment (zeroedin-bill) [inline lib/Doctrine/ORM/Query/Filter/SQLFilter.php]: Block inside this conditional is not reachable.

**Comment by zeroedin-bill** [comment]
@holtkamp Mostly referring to the Traversable possibility. If you pass an array, and the original array changes after it is passed, you would not affect the filter in any way, because the filter would hold a copy. If you pass a Traversable object, and then change that object after it was passed to the filter, the filter's state could change without setting the dirty flag.

I don't think anyone should be passing an object to a SQLfilter period, for that reason. The SQLFilter interface and phpdoc kind of hints that anything passed should be primitive, but it's not specified explicitly. The documentation should be made more explicit about what not to do with a SQLFilter, really, but that's not really in the scope of this PR.

This needs more tests IMO, especially around the ```$type``` param, and with all three types of allowable arrays.

File: lib/Doctrine/ORM/Query/Filter/SQLFilter.php (hunk 1 above)
Comment (beberlei) [reply lib/Doctrine/ORM/Query/Filter/SQLFilter.php:76]: Please revert this change to array inferring, the `ParameterTypeinferer` is already responsible for that itself.

Then this PR is good to go for 2.9.x in my POV.

**Comment by beberlei** [comment]
Closing in favor of #8375

**Pull request closed without merging**